### PR TITLE
Fix autofocusing for long-lived dialog components

### LINKED ISSUE #2397: useDialog autofocus doesn't work for long-lived components
# 🐛 Bug Report 
The `useDialog` hook only attempts to autofocus when the passed in ref changes. Refs generally don't change for the lifetime of a component, so the autofocus only works during the initial component render.

## 🤔 Expected Behavior
Autofocusing should happen when the ref's current value changes. The existing implementation does have a safeguard to prevent stealing focus when focus is already inside the overlay, so it should be safe to run every time the ref changes. This would even handle unusual edge cases where the focus context is lost due to the entire overlay element being replaced in the DOM (e.g. in a multi-screen modal).

## 😯 Current Behavior
When `useDialog` is called for the first time with an empty ref object, it won't autofocus when the content mounts and the ref is hydrated. 

## 💁 Possible Solution
Modifying the deps list in its `useEffect` hook to be `[ref.current]` instead of`[ref]` solves the issue on my end.

## 🔦 Context
I have a dialog component that manages its own presence (the reason for this is to support exit animations, though I can imagine other use cases). It creates a ref to pass to `useDialog` that lives for the duration of the component, which is at least as long as its trigger element.

## 💻 Code Sample
See PR for a test case reproducing the issue.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   | `@react-aria/dialog` v3.1.4
| Browser          | tested in Chrome/Edge, Firefox
| Operating System | Linux, macOS

### PR DESCRIPTION
**Pull request by kherock**
Closes #2397

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

The scenario for `should focus the overlay on open` will fail without this change. For proof, revert
https://github.com/adobe/react-spectrum/blob/6b36d162a6634978740d0f60118c9fb6804850cf/packages/@react-aria/dialog/src/useDialog.ts#L54

and run

```console
$ yarn test packages/@react-aria/dialog
```

## 🧢 Your Project:

N/A

### CODE REVIEW

**Comment by kherock** [comment]
I took a look at the regression caused by this, and I'm doubting that the iOS workaround actually does what it's intended - it seems totally at odds with `shouldCloseOnBlur`. The failing test basically asserts that the tray should close immediately after it's opened!
https://github.com/adobe/react-spectrum/blob/a169a3c827fe07e1e94ff8f7539dd2bbb6874414/packages/@react-spectrum/overlays/test/Tray.test.js#L101-L109

This test seems to be saying "after waiting for the tray to open, it should automatically close due to the iOS workaround". I have a real component implementation resembling `ControlledExample`, and when I enable `shouldCloseOnBlur` without defining an autofocusable element, my dialog does indeed close as soon as I open it.

**Comment by snowystinger** [comment]
Thanks for opening an issue and suggesting a fix 👍🏻 , we'll have to check on all the versions we support to make sure this isn't needed anymore.

File: packages/@react-aria/dialog/src/useDialog.ts [hunk 1]
```diff
@@ -36,22 +36,8 @@ export function useDialog(props: AriaDialogProps, ref: RefObject<HTMLElement>):
...
-      let timeout = setTimeout(() => {
-        if (document.activeElement === ref.current) {
-          ref.current.blur();
-          focusSafely(ref.current);
-        }
-      }, 500);
-
-      return () => {
-        clearTimeout(timeout);
-      };
```
Comment (majornista) [inline packages/@react-aria/dialog/src/useDialog.ts:52]: Why is this hack being removed? Have to tested with Safari on iOS using VoiceOver to confirm that the VoiceOver cursor does indeed get set to a Dialog in all cases?

**Comment by snowystinger** [comment]
Closing for same reason https://github.com/adobe/react-spectrum/issues/2397#issuecomment-1613951081

**Pull request closed without merging**